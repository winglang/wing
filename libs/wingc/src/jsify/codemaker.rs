use indexmap::IndexMap;
use parcel_sourcemap::{OriginalLocation, SourceMap};
use serde_json::json;

use crate::diagnostic::WingSpan;

macro_rules! new_code {
	($source:expr, $($arg:expr),*) => {
		{
			let mut base = CodeMaker::with_source($source);
			$(base.append($arg);)*
			base
		}
	};
}

/// A helper for generating code snippets with indentation.
///
/// TODO: add `open_block` or `close_block` methods that automatically add
/// `{` and `}`?
#[derive(Default, Clone, Debug)]
pub struct CodeMaker {
	lines: Vec<LineData>,
	indent: IndentAmount,
	source: Option<WingSpan>,
}

pub type IndentAmount = usize;
pub type CharacterOffset = usize;

#[derive(Default, Clone, Debug)]
pub struct LineData {
	/// The line of code. Cannot contain any newlines
	pub line: String,
	/// Current indent level
	pub indent: IndentAmount,
	/// Mappings from character offset of this line text to the given source span
	pub mappings: Vec<(CharacterOffset, WingSpan)>,
}

/// Converts stringable data into a codemaker
/// This handles splitting newlines
fn string_to_code<S: Into<String>>(s: S) -> CodeMaker {
	let text: String = s.into();
	let mut code = CodeMaker::default();

	// remove trailing newline
	let line_text = text.strip_suffix("\n").unwrap_or(&text);

	// if the line has newlines in it, consider each line separately
	for subline in line_text.split('\n') {
		code.push_line(LineData {
			line: subline.to_string(),
			indent: 0,
			mappings: vec![],
		});
	}

	code
}

impl From<String> for CodeMaker {
	fn from(s: String) -> Self {
		string_to_code(s)
	}
}

impl From<&String> for CodeMaker {
	fn from(s: &String) -> Self {
		string_to_code(s)
	}
}

impl From<&str> for CodeMaker {
	fn from(s: &str) -> Self {
		string_to_code(s)
	}
}

impl From<Vec<CodeMaker>> for CodeMaker {
	fn from(v: Vec<CodeMaker>) -> Self {
		let mut code: CodeMaker = CodeMaker::default();
		for (idx, line) in v.into_iter().enumerate() {
			if idx > 0 {
				code.append(", ");
			}
			code.append(line);
		}
		code
	}
}

impl CodeMaker {
	pub fn with_source<S: Into<WingSpan>>(source: S) -> Self {
		Self {
			source: Some(source.into()),
			indent: 0,
			lines: vec![],
		}
	}

	/// Emits a line of code and then increases the indent by one.
	pub fn open<S: Into<CodeMaker>>(&mut self, line: S) {
		self.line(line);
		self.indent += 1;
	}

	/// Decreases the indent by one and then emits a line of code.
	pub fn close<S: Into<CodeMaker>>(&mut self, line: S) {
		self.indent -= 1;
		self.line(line);
	}

	/// Emits a line of code with the current indent.
	pub fn line<S: Into<CodeMaker>>(&mut self, line: S) {
		let mut new_code: CodeMaker = line.into();

		for line_data in new_code.lines.drain(..) {
			let mut mappings = line_data.mappings;

			if mappings.is_empty() {
				// if there are no existing mappings, use an existing
				if let Some(source) = &self.source {
					mappings = vec![(0, source.clone())]
				}
			};

			self.push_line(LineData {
				line: line_data.line,
				indent: line_data.indent + self.indent,
				mappings,
			});
		}
	}

	/// Emits an empty line.
	pub fn empty_line(&mut self) {
		self.line("");
	}

	/// Emits multiple lines of code starting with the current indent.
	pub fn add_code(&mut self, code: CodeMaker) {
		assert_eq!(code.indent, 0, "Cannot add code with indent");
		for mut line_data in code.lines {
			line_data.indent += self.indent;
			self.push_line(line_data);
		}
	}

	/// Emits code starting on the current line, mutating it if needed.
	/// Note this mutates the given code.
	pub fn append<S: Into<CodeMaker>>(&mut self, line: S) {
		let mut new_code: CodeMaker = line.into();

		if new_code.is_empty() {
			// nothing to add
			return;
		}

		if self.lines.is_empty() {
			// we need at list one line to append to
			self.empty_line();
		}

		// we know there is at least one line
		let existing_last_line = self.lines.last_mut().expect("last line");

		// Get the first new line and add to to the last line of the current
		let mut new_first_line = new_code.lines.remove(0);
		existing_last_line.line.push_str(&new_first_line.line);

		// Add the mappings from the first line of the new code to the last line of the existing code
		for (offset, span) in new_first_line.mappings.drain(..) {
			existing_last_line
				.mappings
				.push((existing_last_line.line.len() - new_first_line.line.len() + offset, span));
		}

		// Add the remaining lines of the new code (with additional indent)
		for mut line_data in new_code.lines {
			line_data.indent += self.indent;
			self.push_line(line_data);
		}
	}

	/// Decreases the current indent by one.
	#[allow(dead_code)]
	pub fn unindent(&mut self) {
		self.indent -= 1;
	}

	/// Increases the current indent by one.
	pub fn indent(&mut self) {
		self.indent += 1;
	}

	pub fn one_line<S: Into<String>>(s: S) -> CodeMaker {
		let mut code = CodeMaker::default();
		code.line(s.into());
		code
	}

	/// Checks if there are no lines of code
	pub fn is_empty(&self) -> bool {
		self.lines.is_empty()
	}

	fn push_line(&mut self, line: LineData) {
		self.lines.push(line);
	}

	/// Generates a sourcemap for the given source and generated paths, based on the provided source content.
	///
	/// # Remarks
	///
	/// See sourcemap spec for more details: https://sourcemaps.info/spec.html
	///
	/// # Arguments
	///
	/// * `source_path` - Path to the source file, must be the source file reference in the mappings
	/// * `source_content` - Content of the source file
	/// * `generated_path` - Path to the generated (output) file
	///
	/// # Returns
	///
	/// A JSON string that represents the generated sourcemap.
	pub fn generate_sourcemap(&mut self, source_path: &str, source_content: &str, generated_path: &str) -> String {
		let mut sourcemap = SourceMap::new("");
		let source_num = sourcemap.add_source(source_path);

		sourcemap
			.set_source_content(source_num as usize, source_content)
			.expect("Set source content");

		for (line_idx, line) in self.lines.iter().enumerate() {
			let generated_line = line_idx as u32;
			let generated_column = 0;
			if !line.mappings.is_empty() {
				for mapping in &line.mappings {
					let original = OriginalLocation::new(mapping.1.start.line, mapping.1.start.col, 0, None);
					sourcemap.add_mapping(generated_line, mapping.0 as u32, Some(original));
				}
			} else {
				sourcemap.add_mapping(generated_line, generated_column, None);
			}
		}

		let mut buffer: Vec<u8> = vec![];

		sourcemap.write_vlq(&mut buffer).expect("write vlq");

		// Per sourcemap spec, the order of keys is important (version should be first)
		let mut json = IndexMap::new();

		json.insert("version", json!(3));
		json.insert("file", json!(generated_path));
		json.insert("sources", json!(vec![source_path]));
		json.insert("names", json!(sourcemap.get_names()));
		json.insert("mappings", json!(String::from_utf8(buffer).unwrap()));

		serde_json::ser::to_string(&json).expect("serialize sourcemap")
	}
}

impl ToString for CodeMaker {
	fn to_string(&self) -> String {
		let mut code = String::new();
		for (idx, line_data) in self.lines.iter().enumerate() {
			if idx > 0 {
				code.push('\n');
			}
			code.push_str(&"  ".repeat(line_data.indent));
			code.push_str(&line_data.line);
		}
		code
	}
}

#[cfg(test)]
mod tests {
	use crate::diagnostic::WingLocation;

	use super::*;

	use indoc::indoc;
	use serde_json::Value;

	#[test]
	fn codemaker_basic() {
		let mut code = CodeMaker::default();
		code.line("let a = 1;");
		code.open("if (a === 1) {");
		code.line("console.log('a is 1');");
		code.close("}");
		code.line("let b = 2;");
		code.indent();
		code.line("let c = 3;");
		code.unindent();
		code.line("let d = 4;");
		assert_eq!(
			code.to_string(),
			indoc! {r#"
				let a = 1;
				if (a === 1) {
				  console.log('a is 1');
				}
				let b = 2;
				  let c = 3;
				let d = 4;"#}
		);
	}

	#[test]
	fn codemaker_add_code() {
		let mut code1 = CodeMaker::default();
		code1.line("let a = 1;");
		code1.line("let b = 2;");
		let mut code2 = CodeMaker::default();
		code2.open("{");
		code2.add_code(code1);
		code2.close("}");
		assert_eq!(
			code2.to_string(),
			indoc! {r#"
			{
			  let a = 1;
			  let b = 2;
			}"#}
		);
	}

	#[test]
	fn codemaker_line_with_newlines() {
		let mut code = CodeMaker::default();
		code.open("<");
		code.line("hello\nworld");
		code.close(">");
		assert_eq!(
			code.to_string(),
			indoc! {r#"
				<
				  hello
				  world
				>"#}
		);
	}

	#[test]
	fn codemaker_append() {
		let mut code = CodeMaker::default();
		code.append("1");
		code.append("2");
		code.append("\n3");

		assert!(code.lines.len() == 2);
		assert_eq!(code.to_string(), "12\n3");
	}

	#[test]
	fn sourcemap_no_sources() {
		let mut code = CodeMaker::default();

		code.line("1");
		code.line("2");
		code.append("\n3");

		assert!(code.lines.len() == 3);
		assert_eq!(code.to_string(), "1\n2\n3");

		let sourcemap: Value =
			serde_json::from_str(&code.generate_sourcemap("test.w", "1\n2\n3", "test.cjs")).expect("sourcemap");

		// "A" just means the line comes from somewhere in the provided source root
		assert_eq!(sourcemap["mappings"], "A;A;A");
	}

	#[test]
	fn sourcemap_default() {
		let base_span = WingSpan {
			file_id: "test.w".to_string(),
			start: WingLocation { line: 2, col: 4 },
			end: WingLocation { line: 2, col: 8 },
			start_offset: 0,
			end_offset: 0,
		};
		let mut code = CodeMaker::with_source(&base_span);
		code.line("1");
		code.line("2");

		// this line will not inherit the base source
		code.line(CodeMaker::one_line("3"));

		let next_span = WingSpan {
			start: WingLocation { line: 5, col: 10 },
			end: WingLocation { line: 5, col: 20 },
			..base_span
		};
		code.line(new_code!(next_span, "4"));

		assert!(code.lines.len() == 4);
		assert_eq!(code.to_string(), "1\n2\n3\n4");
		let sourcemap: Value =
			serde_json::from_str(&code.generate_sourcemap("test.w", "1\n2\n3", "test.cjs")).expect("sourcemap");

		assert_eq!(sourcemap["version"], 3);
		assert_eq!(sourcemap["file"], "test.cjs");
		assert_eq!(sourcemap["sources"], json!(["test.w"]));
		assert_eq!(sourcemap["names"], json!([]));

		// Useful tool to validate vlq https://www.murzwin.com/base64vlq.html
		assert_eq!(sourcemap["mappings"], "AAEI;AAAA;AAAA;AAGM");
	}

	#[test]
	fn sourcemap_multi_span_line() {
		let base_span = WingSpan {
			file_id: "test.w".to_string(),
			start: WingLocation { line: 2, col: 4 },
			end: WingLocation { line: 2, col: 8 },
			start_offset: 0,
			end_offset: 0,
		};
		let mut code = CodeMaker::with_source(&base_span);
		code.line("1");

		let next_span = WingSpan {
			start: WingLocation { line: 5, col: 10 },
			end: WingLocation { line: 5, col: 20 },
			..base_span
		};
		code.append(new_code!(next_span, "2"));
		code.append("3");
		code.line("4");

		assert!(code.lines.len() == 2);
		assert_eq!(code.to_string(), "123\n4");
		let sourcemap: Value = serde_json::from_str(&code.generate_sourcemap("", "", "")).expect("sourcemap");

		assert_eq!(sourcemap["mappings"], "AAEI,CAGM;AAHN");
	}
}
