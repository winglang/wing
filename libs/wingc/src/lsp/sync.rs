use camino::{Utf8Path, Utf8PathBuf};
use indexmap::IndexMap;
use lsp_types::{DidChangeTextDocumentParams, DidOpenTextDocumentParams};
use wingii::type_system::TypeSystem;

use std::cell::RefCell;
use std::path::{Path, PathBuf};
use tree_sitter::Tree;

use crate::closure_transform::ClosureTransformer;
use crate::diagnostic::{found_errors, reset_diagnostics};
use crate::file_graph::FileGraph;
use crate::files::Files;
use crate::fold::Fold;
use crate::jsify::JSifier;
use crate::lifting::LiftVisitor;
use crate::parser::{normalize_path, parse_wing_project};
use crate::type_check;
use crate::type_check::jsii_importer::JsiiImportSpec;
use crate::type_check::type_reference_transform::TypeReferenceTransformer;
use crate::type_check_assert::TypeCheckAssert;
use crate::valid_json_visitor::ValidJsonVisitor;
use crate::visit::Visit;
use crate::wasm_util::extern_json_fn;
use crate::{ast::Scope, type_check::Types};

/// The output of compiling a Wing project with one or more files
pub struct ProjectData {
	/// Text data contained in the files (utf8)
	pub files: Files,
	/// A graph that tracks the dependencies between files
	pub file_graph: FileGraph,
	/// tree-sitter trees
	pub trees: IndexMap<Utf8PathBuf, Tree>,
	/// AST for each file
	pub asts: IndexMap<Utf8PathBuf, Scope>,
	/// The JSII imports for the file. This is saved so we can load JSII types (for autocompletion for example)
	/// which don't exist explicitly in the source.
	pub jsii_imports: Vec<JsiiImportSpec>,
}

impl ProjectData {
	fn new() -> Self {
		ProjectData {
			files: Files::new(),
			file_graph: FileGraph::default(),
			trees: IndexMap::new(),
			asts: IndexMap::new(),
			jsii_imports: Vec::new(),
		}
	}
}

thread_local! {
	/// When consumed as a WASM library, wingc is not in control of the process/memory in which it is running.
	/// This means that it cannot reliably manage stateful data like this between function calls.
	/// Here we will assume the process is single threaded, and use thread_local to store this data.
	pub static WING_TYPES: RefCell<Types> = RefCell::new(Types::new());
	pub static PROJECT_DATA: RefCell<ProjectData> = RefCell::new(ProjectData::new());
	pub static JSII_TYPES: RefCell<TypeSystem> = RefCell::new(TypeSystem::new());
}

#[no_mangle]
pub unsafe extern "C" fn wingc_on_did_open_text_document(ptr: u32, len: u32) {
	extern_json_fn(ptr, len, on_document_did_open);
}

pub fn on_document_did_open(params: DidOpenTextDocumentParams) {
	WING_TYPES.with(|wing_types| {
		JSII_TYPES.with(|jsii_types| {
			PROJECT_DATA.with(|project_data| {
				let uri = params.text_document.uri;
				let uri_path = uri.to_file_path().unwrap();
				let source_text = params.text_document.text;

				partial_compile(
					&uri_path,
					source_text,
					&mut wing_types.borrow_mut(),
					&mut jsii_types.borrow_mut(),
					&mut project_data.borrow_mut(),
				);
			});
		});
	});
}

#[no_mangle]
pub unsafe extern "C" fn wingc_on_did_change_text_document(ptr: u32, len: u32) {
	extern_json_fn(ptr, len, on_document_did_change);
}

pub fn on_document_did_change(params: DidChangeTextDocumentParams) {
	let DidChangeTextDocumentParams {
		text_document,
		content_changes,
	} = params;

	WING_TYPES.with(|wing_types| {
		JSII_TYPES.with(|jsii_types| {
			PROJECT_DATA.with(|project_data| {
				let uri = text_document.uri;
				let uri_path = uri.to_file_path().unwrap();
				let source_text = content_changes.into_iter().next().unwrap().text;

				partial_compile(
					&uri_path,
					source_text,
					&mut wing_types.borrow_mut(),
					&mut jsii_types.borrow_mut(),
					&mut project_data.borrow_mut(),
				);
			});
		})
	});
}

/// Runs several phases of the wing compiler on a file, including: parsing, type checking, and lifting
/// `ProjectData` is passed with results from previous compilations, and is updated with the results of this compilation.
fn partial_compile(
	source_path: &Path,
	source_text: String,
	mut types: &mut Types, // TODO: does this need to be shared between recompiles?
	jsii_types: &mut TypeSystem,
	project_data: &mut ProjectData,
) {
	// Reset diagnostics before new compilation (`partial_compile` can be called multiple times)
	reset_diagnostics();

	let source_path = Utf8Path::from_path(source_path).expect("invalid unicode path");
	let source_path = normalize_path(source_path, None);

	let topo_sorted_files = parse_wing_project(
		&source_path,
		Some(source_text),
		&mut project_data.files,
		&mut project_data.file_graph,
		&mut project_data.trees,
		&mut project_data.asts,
	);

	// -- DESUGARING PHASE --

	// Transform all inflight closures defined in preflight into single-method resources
	for file in &topo_sorted_files {
		let mut inflight_transformer = ClosureTransformer::new();
		let scope = project_data.asts.remove(file).unwrap();
		let new_scope = inflight_transformer.fold_scope(scope);
		project_data.asts.insert(file.clone(), new_scope);
	}

	// Reset all type information
	types.reset_expr_types();
	types.reset_scope_envs();

	// -- TYPECHECKING PHASE --
	let mut jsii_imports = vec![];

	// Type check all files in topological order (start with files that don't require any other
	// Wing files, then move on to files that depend on those, etc.)
	for file in &topo_sorted_files {
		let mut scope = project_data.asts.remove(file).expect("matching AST not found");
		type_check(
			&mut scope,
			&mut types,
			&file,
			&project_data.file_graph,
			jsii_types,
			&mut jsii_imports,
		);

		// Make sure all type reference are no longer considered references
		let mut tr_transformer = TypeReferenceTransformer { types: &mut types };
		let scope = tr_transformer.fold_scope(scope);

		// Validate the type checker didn't miss anything - see `TypeCheckAssert` for details
		let mut tc_assert = TypeCheckAssert::new(&types, found_errors());
		tc_assert.check(&scope);

		// Validate all Json literals to make sure their values are legal
		let mut json_checker = ValidJsonVisitor::new(&types);
		json_checker.check(&scope);

		project_data.asts.insert(file.clone(), scope);
	}

	// -- LIFTING PHASE --

	let jsifier = JSifier::new(
		&mut types,
		&project_data.files,
		&project_data.file_graph,
		&source_path,
		// out_dir will not be used
		&source_path,
	);
	for file in &topo_sorted_files {
		let mut lift = LiftVisitor::new(&jsifier);
		let scope = project_data.asts.remove(file).expect("matching AST not found");
		lift.visit_scope(&scope);
		project_data.asts.insert(file.clone(), scope);
	}

	// no need to JSify in the LSP
}

pub fn check_utf8(path: PathBuf) -> Utf8PathBuf {
	path.try_into().expect("invalid unicode path")
}

#[cfg(test)]
pub mod test_utils {
	use std::{fs, str::FromStr};
	use uuid::Uuid;

	use lsp_types::*;

	use crate::diagnostic::assert_no_panics;

	use super::on_document_did_open;

	static LANGUAGE_ID: &str = "wing";

	/// Loads a file into the language server
	///
	/// Returns a `TextDocumentPositionParams` with the generated URI and a cursor position of an indicated position in the file (using `//^`)
	///
	/// Example that gives you the position of the "o" in "Resource.":
	/// ```
	/// load_file_with_contents(
	/// r#"
	/// class Resource {
	/// 	static hello() {}
	/// }
	///
	/// Resource.
	///  //^"#
	/// )
	/// ```
	///
	pub fn load_file_with_contents(content: &str) -> TextDocumentPositionParams {
		let temp_dir = tempfile::tempdir().expect("Failed to create temporary directory");
		let filename = format!("{}.main.w", Uuid::new_v4());
		let file_path = temp_dir.path().join(&filename);
		fs::write(&file_path, content).expect("Failed to write to temporary file");
		let file_uri_string = format!("file:///{}", file_path.to_str().unwrap());
		let uri = Url::from_str(&file_uri_string).unwrap();
		on_document_did_open(DidOpenTextDocumentParams {
			text_document: lsp_types::TextDocumentItem {
				uri: uri.clone(),
				language_id: LANGUAGE_ID.to_string(),
				version: 0,
				text: content.to_string(),
			},
		});

		assert_no_panics();

		// find the character cursor position by looking for the character above the ^
		let mut char_pos = 0_i32;
		let mut line_pos = 0_i32;
		// Note: `.chars()` may not 1-to-1 map with characters as expected if using things like emojis
		// This generally acceptable for testing, but should not be assumed to be the case in the real language server
		// TODO: Add support for grapheme clusters
		for char in content.chars() {
			if char == '^' {
				break;
			} else if char == '\n' {
				char_pos = 0;
				line_pos += 1;
			} else {
				char_pos += 1;
			}
		}

		let cursor_position = Position {
			line: (line_pos - 1).max(0) as u32,
			character: char_pos as u32,
		};

		return TextDocumentPositionParams {
			text_document: TextDocumentIdentifier { uri },
			position: cursor_position,
		};
	}

	/// Finds all ranges in the document starting with `//-`
	pub fn get_ranges(content: &str) -> Vec<Range> {
		let lines = content.lines();

		let mut ranges = vec![];
		for line in lines.enumerate() {
			if line.1.contains("//-") {
				let start_col = line.1.match_indices("//-").next().unwrap().0 + 2;
				let end_col = line.1.len();
				let line_num = line.0 - 1;

				ranges.push(Range {
					start: Position {
						line: line_num as u32,
						character: start_col as u32,
					},
					end: Position {
						line: line_num as u32,
						character: end_col as u32,
					},
				});
			}
		}

		ranges
	}
}
