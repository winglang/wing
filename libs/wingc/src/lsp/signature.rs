use itertools::Itertools;
use lsp_types::{
	Documentation, MarkupContent, MarkupKind, ParameterInformation, ParameterLabel, Position, SignatureHelp,
	SignatureInformation,
};

use crate::ast::{CalleeKind, Class, Expr, ExprKind, New, Stmt, StmtKind, Symbol};
use crate::docs::{render_summary, Documented};
use crate::jsify::codemaker::CodeMaker;
use crate::lsp::sync::PROJECT_DATA;
use crate::lsp::sync::WING_TYPES;

use crate::type_check::symbol_env::SymbolEnvRef;
use crate::type_check::{
	resolve_super_method, resolve_user_defined_type, TypeRef, Types, CLASS_INFLIGHT_INIT_NAME, CLASS_INIT_NAME,
};
use crate::visit::{visit_expr, visit_scope, visit_stmt, Visit};
use crate::wasm_util::extern_json_fn;

use super::sync::check_utf8;

#[no_mangle]
pub unsafe extern "C" fn wingc_on_signature_help(ptr: u32, len: u32) -> u64 {
	extern_json_fn(ptr, len, on_signature_help)
}

pub fn on_signature_help(params: lsp_types::SignatureHelpParams) -> Option<SignatureHelp> {
	WING_TYPES.with(|types| {
		let types = types.borrow_mut();
		PROJECT_DATA.with(|project_data| {
			let project_data = project_data.borrow();
			let uri = params.text_document_position_params.text_document.uri;
			let file = check_utf8(uri.to_file_path().ok().expect("LSP only works on real filesystems"));
			let root_scope = &project_data.asts.get(&file).unwrap();

			let mut scope_visitor = ScopeVisitor::new(&types, params.text_document_position_params.position);
			scope_visitor.visit_scope(root_scope);

			let sig_data: (TypeRef, &crate::ast::ArgList) = if scope_visitor.call_stmt.is_some() {
				let stmt = scope_visitor.call_stmt?;
				let class = scope_visitor.class?;

				match &stmt.kind {
					StmtKind::SuperConstructor { arg_list } => {
						if let Some(p) = &class.parent {
							let t = resolve_user_defined_type(&p, &types.get_scope_env(&root_scope), 0).ok()?;
							let init_lookup = t.as_class()?.env.lookup(
								&if t.is_preflight_class() {
									CLASS_INIT_NAME
								} else {
									CLASS_INFLIGHT_INIT_NAME
								}
								.into(),
								None,
							);

							(init_lookup?.as_variable()?.type_, arg_list)
						} else {
							return None;
						}
					}
					_ => return None,
				}
			} else {
				let expr = scope_visitor.call_expr?;
				let env = scope_visitor.call_env?;
				match &expr.kind {
					ExprKind::New(new_expr) => {
						let New { class, arg_list, .. } = new_expr;

						let Some(t) = resolve_user_defined_type(class, &types.get_scope_env(&root_scope), 0).ok() else {
							return None;
						};

						let init_lookup = t.as_class()?.env.lookup(
							&Symbol {
								name: CLASS_INIT_NAME.into(),
								span: Default::default(),
							},
							None,
						);

						(init_lookup?.as_variable()?.type_, arg_list)
					}
					ExprKind::Call { callee, arg_list } => {
						let t = match callee {
							CalleeKind::Expr(expr) => types.get_expr_type(expr),
							CalleeKind::SuperCall(method) => resolve_super_method(method, &env, &types)
								.ok()
								.map_or(types.error(), |t| t.0),
						};

						(t, arg_list)
					}
					_ => return None,
				}
			};

			let sig = sig_data.0.as_function_sig()?;
			let provided_args = sig_data.1;

			let positional_arg_pos = provided_args
				.pos_args
				.iter()
				.enumerate()
				.filter(|(_, arg)| params.text_document_position_params.position <= arg.span.end.into())
				.count();
			let named_arg_pos = provided_args.named_args.iter().find(|arg| {
				arg
					.1
					.span
					.contains_lsp_position(&params.text_document_position_params.position)
			});

			let param_data = sig
				.parameters
				.iter()
				.enumerate()
				.map(|p| {
					if p.0 == sig.parameters.len() - 1 && p.1.typeref.maybe_unwrap_option().is_struct() {
						format!("...{}", p.1.name)
					} else {
						format!("{}: {}", p.1.name, p.1.typeref)
					}
				})
				.collect_vec();

			let active_parameter = if named_arg_pos.is_some() {
				sig.parameters.len() - 1
			} else {
				provided_args.pos_args.len() - positional_arg_pos
			}
			.min(if param_data.is_empty() { 0 } else { param_data.len() - 1 });

			let param_text = param_data.join(", ");
			let label = format!("({}): {}", param_text, sig.return_type);

			let mut sig_docs = CodeMaker::default();
			if let Some(d) = sig_data.0.docs() {
				render_summary(&mut sig_docs, d);
			}
			let sig_docs = sig_docs.to_string();

			let signature_info = SignatureInformation {
				label,
				documentation: Some(Documentation::MarkupContent(MarkupContent {
					kind: MarkupKind::Markdown,
					value: sig_docs.to_string(),
				})),
				parameters: Some(
					sig
						.parameters
						.iter()
						.enumerate()
						.map(|p| {
							let last_arg = p.0 == sig.parameters.len() - 1;
							let p_type = p.1.typeref;
							let structy = p_type.maybe_unwrap_option();
							let structy = structy.as_struct();

							let docstring = p.1.docs.render();
							let p_docs = if docstring.is_empty() {
								None
							} else {
								Some(Documentation::MarkupContent(MarkupContent {
									kind: MarkupKind::Markdown,
									value: if sig_docs.is_empty() {
										docstring
									} else {
										format!("{docstring}\n\n")
									},
								}))
							};

							ParameterInformation {
								label: if last_arg && structy.is_some() {
									ParameterLabel::Simple(format!("...{}", p.1.name))
								} else {
									ParameterLabel::Simple(param_data.get(p.0).unwrap_or(&format!("{}: {}", p.0, p_type)).clone())
								},
								documentation: if structy.is_some() {
									//check if this is the last arg, allowing for expansion syntax
									if p.0 == sig.parameters.len() - 1 {
										Some(Documentation::MarkupContent(MarkupContent {
											kind: MarkupKind::Markdown,
											value: format!("```wing\n{}\n```\n", p_type.render_docs()),
										}))
									} else {
										p_docs
									}
								} else {
									p_docs
								},
							}
						})
						.collect(),
				),

				active_parameter: Some(active_parameter as u32),
			};

			Some(SignatureHelp {
				signatures: vec![signature_info],
				active_signature: None,
				active_parameter: None,
			})
		})
	})
}

/// This visitor is used to find the scope
/// and relevant expression that contains a given location.
pub struct ScopeVisitor<'a> {
	types: &'a Types,
	/// The target location we're looking for
	pub location: Position,
	/// The nearest expression before (or containing) the target location
	pub call_expr: Option<&'a Expr>,
	/// The nearest statement before (or containing) the target location - on a case we're on a super() call
	pub call_stmt: Option<&'a Stmt>,
	// The env of the found expression
	pub call_env: Option<SymbolEnvRef>,
	/// The current symbol env we're in
	curr_env: Vec<SymbolEnvRef>,
	/// the nearest class containing the call_stmt
	pub class: Option<&'a Class>,
}

impl<'a> ScopeVisitor<'a> {
	pub fn new(types: &'a Types, location: Position) -> Self {
		Self {
			types,
			location,
			call_expr: None,
			call_stmt: None,
			call_env: None,
			class: None,
			curr_env: vec![],
		}
	}
}

impl<'a> Visit<'a> for ScopeVisitor<'a> {
	fn visit_expr(&mut self, node: &'a Expr) {
		visit_expr(self, node);

		if self.call_expr.is_some() || self.call_stmt.is_some() {
			return;
		}

		if node.span.contains_lsp_position(&self.location) {
			match node.kind {
				ExprKind::Call { .. } | ExprKind::New { .. } => {
					self.call_expr = Some(node);
					self.call_env = Some(*self.curr_env.last().unwrap());
				}
				_ => {}
			}
		}
	}

	fn visit_stmt(&mut self, node: &'a Stmt) {
		visit_stmt(self, node);

		if self.call_expr.is_some() || (self.call_stmt.is_some() && self.class.is_some()) {
			return;
		}

		if node.span.contains_lsp_position(&self.location) {
			match &node.kind {
				StmtKind::SuperConstructor { .. } => {
					self.call_stmt = Some(node);
				}
				StmtKind::Class(c) => {
					self.class = Some(c);
				}
				_ => {}
			}
		}
	}

	fn visit_scope(&mut self, node: &'a crate::ast::Scope) {
		let env = self.types.get_scope_env(node);
		self.curr_env.push(env);
		visit_scope(self, node);
		self.curr_env.pop();
	}
}

#[cfg(test)]
mod tests {
	use crate::lsp::signature::*;
	use crate::lsp::sync::test_utils::*;
	use lsp_types::*;

	/// Creates a snapshot test for a given wing program's signature at a given position
	/// In the wing program, place a comment "//^" into the text where the "^" is pointing to the desired character position
	///
	/// First parameter will be the name of the tests, as well as the identifier to use for the list of completion in the asserts (see last parameter)
	/// Second parameter is the wing code block as a string literal
	/// After the first two parameters, any additional are optional statements that should be used for asserting on the given completions.
	///
	/// Result is a list of [CompletionItem]s
	macro_rules! test_signature {
		($name:ident, $code:literal, $($assertion:stmt)*) => {
			#[test]
			fn $name() {
				let text_document_position = load_file_with_contents($code);
				let sig = on_signature_help(SignatureHelpParams {
					context: None,
					text_document_position_params: TextDocumentPositionParams {
						text_document: text_document_position.text_document.clone(),
						position: Position {
							line: text_document_position.position.line,
							character: text_document_position.position.character,
						},
					},
					work_done_progress_params: Default::default(),
				});

				if let Some($name) = sig {
					insta::with_settings!(
						{
							prepend_module_to_snapshot => false,
							omit_expression => true,
							snapshot_path => "./snapshots/signature",
						}, {
							insta::assert_yaml_snapshot!($name);
						}
					);
					$($assertion)*
				}
			}
		};
	}

	test_signature!(empty, "", assert!(false));

	test_signature!(
		second_arg_active,
		r#"
bring cloud;
let bucket = new cloud.Bucket();
bucket.addObject("key", )
                     //^
"#,
	);

	test_signature!(
		named_arg_active,
		r#"
bring cloud;
let bucket = new cloud.Bucket();
bucket.onEvent(inflight () => {
  bucket.delete("", );
                 //^
});
"#,
	);

	test_signature!(
		constructor_arg,
		r#"
bring cloud;
let bucket = new cloud.Bucket( );
                            //^
"#,
	);

	test_signature!(
		nested_class_calls,
		r#"
bring http;

class T {
  inflight do() {
    this.handle(inflight (context: str): str => {
      http.get( )
             //^
      return "321";
    });
  }

  inflight handle(handler:  (str): str) {
    handler("123");
  }
}
"#,
	);

	test_signature!(
		class_super,
		r#"
    class Base {
      new(hello: num) {}
    }
    
    class Derived extends Base {
      new() {
        super( );
            //^
      }
    }
    
    new Derived();
"#,
	);

	test_signature!(
		inflight_class_super,
		r#"
    inflight class A {
      pub sound: str;
    
      inflight new(sound: str) {
        this.sound = sound;
      }
    }
    
    inflight class B extends A {
      inflight new(sound: str) {
        super( );
            //^
      }
    }
"#,
	);

	test_signature!(
		empty_super,
		r#"
    class A {}
    
    class B extends A {
      new() {
        super( );
            //^
      }
    }
"#,
	);
}
