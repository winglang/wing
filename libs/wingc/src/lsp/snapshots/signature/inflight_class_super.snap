---
source: libs/wingc/src/lsp/signature.rs
---
signatures:
  - label: "(sound: str): A"
    documentation:
      kind: markdown
      value: "### Parameters\n- `sound` â€” `str`"
    parameters:
      - label: "sound: str"
    activeParameter: 0

